Source/EasyNetQ/PersistentConnection.cs

```diff
@@ -100,6 +102,30 @@ public void Dispose()
             }
         }
 
+        private IModel GetModel(IAutorecoveringConnection connection)
+        {
+            lock (mutex)
+            {
+                if (connection.ChannelMax == 0)
+                {
+                    // unlimited channels available: no need to manage channels
+                    return connection.CreateModel();
+                }
+                else if (models.Count < connection.ChannelMax)
+                {
+                    var model = connection.CreateModel();
+                    model.ModelShutdown += (sender, args) => models.Remove(model);
+                    models.Add(model);
+
+                    return model;
+                }
+                else
+                {
+                    return models.Last();
```

Comment:
I don't now if returning the last one is the right approach, as you said you know that this is not perfect, but I would also recommend to push something like this on production, you could end up with a channel that as a lot of subscription with all the others with just one.

Maybe a simple randomizer could do the trick? or a round robin using a simple counter (it is already thread safe this part)

Reply:
Yes, I thought about it. For my use case (only use 1 channel at all) it works. I did not implement any other options since I wasn't sure about the best option. Maybe it is intended to have one connection for each queue by the dev and to limit side effects it would be best to only pollute one of the connections. As I said, I don't know the right answer to this.

Reply:
From https://www.rabbitmq.com/channels.html#channel-max: 
> Should the configured limit be exceeded, the connection will be closed with a fatal error:
> 2019-02-11 16:04:06.296 [error] <0.887.0> Error on AMQP connection <0.887.0> (127.0.0.1:49956 -> 127.0.0.1:5672, vhost: '/', user: 'guest', state: running), channel 23:
 operation none caused a connection exception not_allowed: "number of channels opened (22) has reached the negotiated channel_max (22)"

It seems to me that the list of connections does not make sense here. Why is the last created connection returned? Why not the first? These connections are used by someone if they are on this list, so it is most likely that the application will work with bugs. Using the `RequestedChannelMax` option, the number of channels is already limited. Good. This will not allow the client intentionally or mistaken to create them more than it follows. Here is not the place to try to solve the problem of channel pooling. At least I do not understand how the channel pooling is related to the `RequestedChannelMax` option. These are independent features.

Reply:
we can also just throw an exception. I don't mind. what do you propose @sungam3r ?

Reply:
I propose to use underlying rabbit-dotnet-client behavior. No need in GetModel method at all. With the new max channel setting we already have throw behavior out of the box.

Reply:
I can't follow. Can you make an example?

Reply:
Example of what? Just revert this file, that's all.

Reply:
Reverting this file will make the hole pull request pointless. The max channel options wont be applied and I can't limit the created channels. If it would work the only thing it would do is throwing an exception (when max channel is reached)  instead of working with a limited number of channels. Then there would be no point to set this option in the first place.

Reply:
> The max channel options wont be applied 

Why? It's rabbit-dotnet-client option.

> If it would work the only thing it would do is throwing an exception (when max channel is reached) instead of working with a limited number of channels.

Exactly. The point is to just allow to set this setting.

> Then there would be no point to set this option in the first place.

The point is to protect against excessive channels growth.

If your goal is to work with a limited number of channels then it's some sort of channel pooling and IMHO a separate issue although I doubt that this makes sense,  because it contradicts the design of channels.

Reply:
The option to set max channel is crucial, like the rabbitmq-client-dotnet. I have more than 2000 channels in some configuration and EasyNetQ become useless if I cannot set it higher.

Reply:
@sungam3r since different services can limit the available amount of channels we (I) need channel pooling. Whats the point in setting the max channel option if it will fail (throw an exception) anyways? As a user of the framework I would hope that the framework takes care of handling channels. When the framework creates to many for the endpoint it should be able to decrease it. I also believe that there might be something wrong when the framework creates 2000 or more channels like @LACS2 describes 👀

Reply:
I understand what you want. I am trying to explain why it will most likely work badly in real scenarios. So, now the introduction of new settings works simultaneously to the restriction of max channels at underlying rabbit-dotnet-client layer, and to use channel "pooling". Moreover, as @micdenny said the "pooling" is implemented primitive, returning always the last channel. Of course this channel is used, because it is in the list, it means it was not closed. All channels in the list are not closed. What will happen when one channel begins to be used for completely different operations? Most likely, the client will receive UNEXPECTED_FRAME error, see for example https://github.com/rabbitmq/rabbitmq-dotnet-client/issues/800. If you need channel pooling, then it should be made on another level of abstraction and taking into account the actual use of these channels. I repeat that I do not have anything against the introduction of a new `RequestedChannelMax` setting as is and passing it to the underlying `ConnectionFactory`.